### PR TITLE
lxd/instance/drivers: Fix raw.lxc handling for shutdown/stop

### PR DESCRIPTION
**Pull request by monstermunchkin**
Due to a rework of the raw.lxc handling, the config wouldn't be loaded
when stopping or shutting down a container.  This caused some keys,
e.g. `lxc.signal.*`, to be ignored.

Signed-off-by: Thomas Hipp <thomas.hipp@canonical.com>

### CODE REVIEW

**Comment by monstermunchkin** [comment]
This behavior was introduced in da7ff328bb4ec8d07b944c413c929cb606d147f0.

File: lxd/instance/drivers/driver_lxc.go [hunk 1]
```diff
@@ -2684,12 +2684,32 @@ func (d *lxc) Stop(stateful bool) error {
...
+
+		err = shared.WriteAll(f, []byte(lxcConfig))
+		f.Close()
+		defer os.Remove(f.Name())
+		if err != nil {
+			return err
+		}
+
+		// Load the config.
+		err = d.c.LoadConfigFile(f.Name())
```
Comment (tomponline) [inline lxd/instance/drivers/driver_lxc.go:2709]: @monstermunchkin why cant we load the config file that is generated at start time and written to the log dir?

**Comment by tomponline** [comment]
@tomponline actually looking at https://github.com/lxc/lxd/commit/da7ff328bb4ec8d07b944c413c929cb606d147f0 more closely, it only writes the raw.lxc config to a temporary file which is only loaded into a lib lxc instance for the start action.

Are the `lxc.signal` entries you're needing existing in the main config file or only in `raw.lxc`? If the latter, then I don't suppose my suggested change actually works, and instead you do infact need to write a temporary file containing `raw.lxc` and then load it into the liblxc instance. However as we are doing that in a few places now, moving that to its own function as part of the lxc instance struct would seem to make sense.

**Comment by monstermunchkin** [comment]
> Are the `lxc.signal` entries you're needing existing in the main config file or only in `raw.lxc`?

They are in `raw.lxc` only.

> If the latter, then I don't suppose my suggested change actually works,

Actually, it does :slightly_smiling_face:

**Comment by tomponline** [comment]
> > Are the `lxc.signal` entries you're needing existing in the main config file or only in `raw.lxc`?
> 
> They are in `raw.lxc` only.
> 
> > If the latter, then I don't suppose my suggested change actually works,
> 
> Actually, it does slightly_smiling_face

Oh you know why, its because the config that is written to the temp file is loaded into the liblxc instance and then written out to the final complete config file. :)

File: lxd/instance/drivers/driver_lxc.go [hunk 2]
```diff
@@ -2690,6 +2690,12 @@ func (d *lxc) Stop(stateful bool) error {
 			op.Done(err)
 			return err
 		}
+
+		// Load the config.
+		err = d.c.LoadConfigFile(filepath.Join(d.LogPath(), "lxc.conf"))
+		if err != nil {
+			return fmt.Errorf("Failed to load raw.lxc")
```
Comment (tomponline) [inline lxd/instance/drivers/driver_lxc.go:2697]: Please could you improve the error to include the error message and the path of the file being loaded?

Comment (tomponline) [reply lxd/instance/drivers/driver_lxc.go:2697]: Also adding a comment as to why this is needed would be good for future reference, something along the lines that certain config keys are used at instance stop time (such as `lxc.signal`).

**Review by tomponline — APPROVED** [review]

**Comment by monstermunchkin** [comment]
All done

**Comment by stgraber** [comment]
LGTM, will merge once jenkins has a chance to test it

**Comment by stgraber** [comment]
Looking rather red :)

**Comment by monstermunchkin** [comment]
@tomponline I had to change some things. The problem was that we only needed `raw.lxc`, and none of the other config parts. By loading the conf from log dir, we'd end up with all the config which caused failures.

**Review by tomponline — APPROVED** [review]

**Comment by tomponline** [comment]
That makes sense, and the code looks good. By moving it to a separate function we can easily update that in the future of the go lxc binding ever gets the ability to load config from a string/io.Reader rather than a file name.

**Pull request merged**